content/6_Advanced/Wavelet.mdx

```diff
@@ -32,9 +30,232 @@ Like a segment tree on values rather than indices.
+</Info>
+
+Therefore, the occurrences of the left node is
+
+$$
+\texttt{left\_occurrences} = prefixB[r] - prefixB[l]
+$$
+
+<Info title="Left occurrences">
+  Note that $\texttt{left\_occurrences}$ is the number of indices between l and r whose value is less than mid
+
+</Info>
+
+- If $\texttt{left\_occurrences}$ is greater or equal to $k$, it means the $k$-th smallest element is in
+  the left subtree. Therefore, we update our range and recurse into the left
+  child
+- If $\texttt{left\_occurrences}$ is less than $k$, it means the
+  $k$-th smallest element is in the right subtree. We adjust k by subtracting
+  $\texttt{left\_occurrences}$ from $k$, update our range, and recurse into the right child
+
+<Info title="Notice">
+  Notice we still update $l, r$ accordingly when we go left or right
+</Info>
+
+the answer then will be the value of the node we end up on (leaf)
+
+## Implemention
+**Time Complexity:** $\mathcal{O}(Q \cdot \log(M))$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+struct Segment {
+	Segment *left = nullptr, *right = nullptr;
+	int l, r, mid;
+	bool children = false;
+	vector<pair<int, int>> indices;  // Index, Value
+	vector<int> prefixB;
+
+	Segment(int l, int r, const vector<pair<int, int>> &indices)
+	    : l(l), r(r), mid((r + l) / 2), indices(indices) {
+		calculate_b();
+	}
+
+	// Sparse since values can go up to 1e9
+	void update() {
+		if (children) { return; }
+		children = true;
+		if (r - l > 1) {
+			// Split the indices for left and right child
+			vector<pair<int, int>> leftIndices, rightIndices;
+			partition_copy(indices.begin(), indices.end(), leftIndices.begin(),
+			               rightIndices.begin(), [this](const pair<int, int> &elem) {
+				               return elem.second < mid;
+			               });
+
+			left = new Segment(l, mid, leftIndices);
+			right = new Segment(mid, r, rightIndices);
+		}
+	}
+
+	// Calculates the prefix B
+	void calculate_b() {
+		int i = 1;
+		int j = 0;
+		prefixB.resize(indices.size() + 1);
+		for (auto [ind, val] : indices) {
+			if (val < mid) j++;
+			prefixB[i++] = j;
+		}
+	}
+
+	int find_k_smallest(int a, int b, int k) {
+		update();
+		if (r - l <= 1) { return l; }
+
+		int lb = prefixB[a];
+		int lr = prefixB[b];
+		int inLeft = lr - lb;  // Amount of values in range (a,b) that are less the mid
+
+		if (k <= inLeft) {
+			return left->find_k_smallest(lb, lr, k);  // Appears in left
+		} else {
+			return right->find_k_smallest(a - lb, b - lr,
+			                              k - inLeft);  // Appears in right
+		}
+	}
+};
+
+int main() {
+	int n, q;
+	cin >> n >> q;
+
+	vector<pair<int, int>> indices;
+	for (int i = 0; i < n; ++i) {
+		int v;
+		cin >> v;
+		indices.emplace_back(i, v);
+	}
+	Segment seg(0, 1e9 + 2, indices);
```

Comment:
just curious but why `1e9 + 2`

Reply:
The main reason is I am afraid of bouncing issues so I added a random number, technically +1 would have worked just as well.

Reply:
could you define this as a constant? `MAX_VAL` or similar

Reply:
Ye that is probably best.